### PR TITLE
 LazyTensors can broadcast matmuls

### PR DESCRIPTION
**Pull request by gpleiss**
This applies to all LazyTensors except: `BatchRepeatLazyTensor`, `BlockDiagonalLazyTensor`,`SumBatchLazyTensor`.

### CODE REVIEW

File: gpytorch/lazy/constant_mul_lazy_tensor.py [hunk 1]
```diff
@@ -113,19 +107,17 @@ def _size(self):
...
     @property
-    def constant(self):
+    def expanded_constant(self):
         # Make sure that the constant can be expanded to the appropriate size
         try:
-            constant = self._constant.expand(self.base_lazy_tensor.batch_shape)
+            constant = self._constant.view(*self._constant.shape, 1, 1)
         except RuntimeError:
             raise RuntimeError(
                 "ConstantMulLazyTensor of size {} received an invalid constant of size {}.".format(
```
Comment (Balandat) [inline gpytorch/lazy/constant_mul_lazy_tensor.py]: Consistency nit: below you format as `... ({}) ...`, let's be consistent in that regard.

File: gpytorch/lazy/interpolated_lazy_tensor.py [hunk 2]
```diff
@@ -164,39 +157,38 @@ def _quad_form_derivative(self, left_vecs, right_vecs):
...
+        flattened_right_interp_right_res = right_interp_right_res.view(batch_size * n_inducing, n_vecs)
+        selected_right_vals = flattened_right_interp_right_res.index_select(0, batched_left_interp_indices)
+        selected_right_vals = selected_right_vals.view(*batch_shape, n_left_rows, n_left_interp, n_vecs)
         left_values_grad = (selected_right_vals * left_vecs.unsqueeze(-2)).sum(-1)
 
         # right_interp_values_grad
         left_interp_left_res = self.base_lazy_tensor._t_matmul(left_res).contiguous()
-        if self.right_interp_indices.ndimension() == 3:
-            batch_offset = torch.arange(0, batch_size, dtype=torch.long, device=self.device)
+        batch_shape = torch.Size(left_interp_left_res.shape[:-2])
```
Comment (Balandat) [inline gpytorch/lazy/interpolated_lazy_tensor.py]: this call to `Size` is extraneous

File: gpytorch/lazy/interpolated_lazy_tensor.py [hunk 3]
```diff
@@ -438,16 +408,11 @@ def sum_batch(self, sum_batch_size=None):
...
-            res = left_interp(
-                self.left_interp_indices, self.left_interp_values, base_samples.permute(1, 2, 0).contiguous()
-            )
-            return res.permute(2, 0, 1).contiguous()
-        else:
-            res = left_interp(
-                self.left_interp_indices, self.left_interp_values, base_samples.permute(1, 0).contiguous()
-            )
-            return res.permute(1, 0).contiguous()
+        batch_iter = tuple((i + 1) for i in range(base_samples.dim() - 1))
```
Comment (Balandat) [inline gpytorch/lazy/interpolated_lazy_tensor.py]: `tuple(range(base_samples.dim()))`

File: gpytorch/lazy/interpolated_lazy_tensor.py [hunk 4]
```diff
@@ -438,16 +408,11 @@ def sum_batch(self, sum_batch_size=None):
...
+        base_samples = base_samples.permute(*batch_iter, 0)
+        res = left_interp(self.left_interp_indices, self.left_interp_values, base_samples).contiguous()
+        batch_iter = tuple(i for i in range(res.dim() - 1))
```
Comment (Balandat) [inline gpytorch/lazy/interpolated_lazy_tensor.py]: `tuple(range(res.dim() - 1))`

File: gpytorch/lazy/mul_lazy_tensor.py [hunk 5]
```diff
@@ -106,14 +107,16 @@ def _args(self, args):
         pass
 
     def _matmul(self, rhs):
+        output_shape = _matmul_broadcast_shape(self.shape, rhs.shape)
+        output_batch_shape = torch.Size(output_shape[:-2])
```
Comment (Balandat) [inline gpytorch/lazy/mul_lazy_tensor.py]: extraneous `Size` call

File: gpytorch/lazy/kronecker_product_lazy_tensor.py [hunk 6]
```diff
@@ -4,32 +4,41 @@
...
 
 def _prod(iterable):
     return reduce(operator.mul, iterable, 1)
 
 
-def _matmul(lazy_tensors, rhs):
-    res = rhs.contiguous()
+def _matmul(lazy_tensors, kp_shape, rhs):
+    output_shape = _matmul_broadcast_shape(kp_shape, rhs.shape)
+    output_batch_shape = torch.Size(output_shape[:-2])
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_lazy_tensor.py]: extraneous `Size` call

File: gpytorch/lazy/kronecker_product_lazy_tensor.py [hunk 7]
```diff
@@ -4,32 +4,41 @@
...
+        res = factor.contiguous().view(*output_batch_shape, -1, num_cols)
     return res
 
 
-def _t_matmul(lazy_tensors, rhs):
-    res = rhs.contiguous()
+def _t_matmul(lazy_tensors, kp_shape, rhs):
+    kp_t_shape = torch.Size((*kp_shape[:-2], kp_shape[-1], kp_shape[-2]))
+    output_shape = _matmul_broadcast_shape(kp_t_shape, rhs.shape)
+    output_batch_shape = torch.Size(output_shape[:-2])
```
Comment (Balandat) [inline gpytorch/lazy/kronecker_product_lazy_tensor.py]: extraneous `Size` call

File: gpytorch/utils/sparse.py [hunk 8]
```diff
@@ -84,10 +83,31 @@ def bdsmm(sparse, dense):
...
     Batch dense-sparse matrix multiply
     """
+    # Make the batch sparse matrix into a block-diagonal matrix
     if sparse.ndimension() > 2:
-        batch_size, num_rows, num_cols = sparse.size()
-        batch_assignment = sparse._indices()[0]
-        indices = sparse._indices()[1:].clone()
+        # Expand the tensors to account for broadcasting
+        output_shape = _matmul_broadcast_shape(sparse.shape, dense.shape)
+        expanded_sparse_shape = torch.Size(output_shape[:-2] + sparse.shape[-2:])
```
Comment (Balandat) [inline gpytorch/utils/sparse.py]: extraneous `Size` call

File: gpytorch/utils/toeplitz.py [hunk 9]
```diff
@@ -215,34 +185,21 @@ def sym_toeplitz_derivative_quadratic_form(left_vectors, right_vectors):
         left_vectors = left_vectors.unsqueeze(1)
         right_vectors = right_vectors.unsqueeze(1)
 
-    left_vectors = left_vectors.transpose(-1, -2)
-    right_vectors = right_vectors.transpose(-1, -2)
+    batch_shape = torch.Size(left_vectors.shape[:-2])
```
Comment (Balandat) [inline gpytorch/utils/toeplitz.py]: extraneous `Size` call

File: gpytorch/utils/sparse.py [hunk 10]
```diff
@@ -84,10 +83,31 @@ def bdsmm(sparse, dense):
...
+
+        # Figure out how much need to be added to the row/column indices to create
+        # a block-diagonal matrix
+        *batch_shape, num_rows, num_cols = sparse.shape
+        batch_size = torch.Size(batch_shape).numel()
+        batch_multiplication_factor = torch.tensor(
+            [torch.Size(batch_shape[i + 1:]).numel() for i in range(len(batch_shape))],
+            dtype=torch.long, device=sparse.device
+        )
+        batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
```
Comment (Balandat) [inline gpytorch/utils/sparse.py]: on the GPU, this errors out with the following: 
```
[...]

/data/users/balandat/fbsource/fbcode/buck-out/dev/gen/bento/kernels/bento_kernel_ae_lazarus#link-tree/gpytorch/utils/sparse.py in bdsmm(sparse, dense)
    105             dtype=torch.long, device=sparse.device
    106         )
--> 107         batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
    108 
    109         # Create block-diagonal sparse tensor

RuntimeError: addmv for CUDA tensors only supports floating-point types. Try converting the tensors with .float() at caffe2/aten/src/THC/generic/THCTensorMathBlas.cu:150
```

You can either do all of this index business on the cpu, or just add a little if statement in to perform the computation as float on the GPU and then cast back to long (I'm sure there is a more elegant way of doing this...):
```
if sparse.is_cuda:
    batch_assignment = (sparse._indices()[:-2].t().float() @ batch_multiplication_factor.float()).to(dtype=torch.long)
else:
    batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
```

Comment (gpleiss) [reply gpytorch/utils/sparse.py]: Whoops sorry my comment was stupid. Doing it your way 😁

**Pull request merged**